{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "74834865",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: discord.py in c:\\users\\omarm\\anaconda3\\lib\\site-packages (2.2.2)\n",
      "Requirement already satisfied: aiohttp<4,>=3.7.4 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from discord.py) (3.8.4)\n",
      "Requirement already satisfied: attrs>=17.3.0 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (21.4.0)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (1.3.3)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (6.0.4)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (1.8.2)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (2.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (4.0.2)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from aiohttp<4,>=3.7.4->discord.py) (1.3.1)\n",
      "Requirement already satisfied: idna>=2.0 in c:\\users\\omarm\\anaconda3\\lib\\site-packages (from yarl<2.0,>=1.0->aiohttp<4,>=3.7.4->discord.py) (3.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install -U discord.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9fbebb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2023-03-16 11:46:36] [INFO    ] discord.client: logging in using static token\n",
      "[2023-03-16 11:46:36] [INFO    ] discord.client: logging in using static token\n",
      "[2023-03-16 11:46:36] [INFO    ] discord.client: logging in using static token\n",
      "[2023-03-16 11:46:36] [INFO    ] discord.client: logging in using static token\n",
      "[2023-03-16 11:46:36] [INFO    ] discord.client: logging in using static token\n",
      "[2023-03-16 11:46:37] [INFO    ] discord.gateway: Shard ID None has connected to Gateway (Session ID: 2d38f39fd1016ff911dbfdd3591f4eca).\n",
      "[2023-03-16 11:46:37] [INFO    ] discord.gateway: Shard ID None has connected to Gateway (Session ID: 2d38f39fd1016ff911dbfdd3591f4eca).\n",
      "[2023-03-16 11:46:37] [INFO    ] discord.gateway: Shard ID None has connected to Gateway (Session ID: 2d38f39fd1016ff911dbfdd3591f4eca).\n",
      "[2023-03-16 11:46:37] [INFO    ] discord.gateway: Shard ID None has connected to Gateway (Session ID: 2d38f39fd1016ff911dbfdd3591f4eca).\n",
      "[2023-03-16 11:46:37] [INFO    ] discord.gateway: Shard ID None has connected to Gateway (Session ID: 2d38f39fd1016ff911dbfdd3591f4eca).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logged in as Final_Bot#0402 (ID: 1085927648431190077)\n",
      "------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2023-03-16 11:49:12] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command None\n",
      "discord.ext.commands.errors.CommandNotFound: Command \"hello\" is not found\n",
      "[2023-03-16 11:49:12] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command None\n",
      "discord.ext.commands.errors.CommandNotFound: Command \"hello\" is not found\n",
      "[2023-03-16 11:49:12] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command None\n",
      "discord.ext.commands.errors.CommandNotFound: Command \"hello\" is not found\n",
      "[2023-03-16 11:49:12] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command None\n",
      "discord.ext.commands.errors.CommandNotFound: Command \"hello\" is not found\n",
      "[2023-03-16 11:49:12] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command None\n",
      "discord.ext.commands.errors.CommandNotFound: Command \"hello\" is not found\n",
      "[2023-03-16 11:53:50] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command roll\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\bot.py\", line 1350, in invoke\n",
      "    await ctx.command.invoke(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 1015, in invoke\n",
      "    await self.prepare(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 932, in prepare\n",
      "    await self._parse_arguments(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 839, in _parse_arguments\n",
      "    transformed = await self.transform(ctx, param, attachments)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 691, in transform\n",
      "    raise MissingRequiredArgument(param)\n",
      "discord.ext.commands.errors.MissingRequiredArgument: dice is a required argument that is missing.\n",
      "[2023-03-16 11:53:50] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command roll\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\bot.py\", line 1350, in invoke\n",
      "    await ctx.command.invoke(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 1015, in invoke\n",
      "    await self.prepare(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 932, in prepare\n",
      "    await self._parse_arguments(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 839, in _parse_arguments\n",
      "    transformed = await self.transform(ctx, param, attachments)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 691, in transform\n",
      "    raise MissingRequiredArgument(param)\n",
      "discord.ext.commands.errors.MissingRequiredArgument: dice is a required argument that is missing.\n",
      "[2023-03-16 11:53:50] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command roll\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\bot.py\", line 1350, in invoke\n",
      "    await ctx.command.invoke(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 1015, in invoke\n",
      "    await self.prepare(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 932, in prepare\n",
      "    await self._parse_arguments(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 839, in _parse_arguments\n",
      "    transformed = await self.transform(ctx, param, attachments)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 691, in transform\n",
      "    raise MissingRequiredArgument(param)\n",
      "discord.ext.commands.errors.MissingRequiredArgument: dice is a required argument that is missing.\n",
      "[2023-03-16 11:53:50] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command roll\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\bot.py\", line 1350, in invoke\n",
      "    await ctx.command.invoke(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 1015, in invoke\n",
      "    await self.prepare(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 932, in prepare\n",
      "    await self._parse_arguments(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 839, in _parse_arguments\n",
      "    transformed = await self.transform(ctx, param, attachments)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 691, in transform\n",
      "    raise MissingRequiredArgument(param)\n",
      "discord.ext.commands.errors.MissingRequiredArgument: dice is a required argument that is missing.\n",
      "[2023-03-16 11:53:50] [ERROR   ] discord.ext.commands.bot: Ignoring exception in command roll\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\bot.py\", line 1350, in invoke\n",
      "    await ctx.command.invoke(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 1015, in invoke\n",
      "    await self.prepare(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 932, in prepare\n",
      "    await self._parse_arguments(ctx)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 839, in _parse_arguments\n",
      "    transformed = await self.transform(ctx, param, attachments)\n",
      "  File \"C:\\Users\\omarm\\anaconda3\\lib\\site-packages\\discord\\ext\\commands\\core.py\", line 691, in transform\n",
      "    raise MissingRequiredArgument(param)\n",
      "discord.ext.commands.errors.MissingRequiredArgument: dice is a required argument that is missing.\n"
     ]
    }
   ],
   "source": [
    "# This example requires the 'members' and 'message_content' privileged intents to function.\n",
    "\n",
    "import discord\n",
    "from discord.ext import commands\n",
    "import random\n",
    "import nest_asyncio\n",
    "nest_asyncio.apply()\n",
    "\n",
    "\n",
    "description = '''An example bot to showcase the discord.ext.commands extension\n",
    "module.\n",
    "\n",
    "There are a number of utility commands being showcased here.'''\n",
    "\n",
    "intents = discord.Intents.default()\n",
    "intents.members = True\n",
    "intents.message_content = True\n",
    "\n",
    "bot = commands.Bot(command_prefix='?', description=description, intents=intents)\n",
    "\n",
    "\n",
    "@bot.event\n",
    "async def on_ready():\n",
    "    print(f'Logged in as {bot.user} (ID: {bot.user.id})')\n",
    "    print('------')\n",
    "\n",
    "\n",
    "@bot.command()\n",
    "async def add(ctx, left: int, right: int):\n",
    "    \"\"\"Adds two numbers together.\"\"\"\n",
    "    await ctx.send(left + right)\n",
    "\n",
    "\n",
    "@bot.command()\n",
    "async def roll(ctx, dice: str):\n",
    "    \"\"\"Rolls a dice in NdN format.\"\"\"\n",
    "    try:\n",
    "        rolls, limit = map(int, dice.split('d'))\n",
    "    except Exception:\n",
    "        await ctx.send('Format has to be in NdN!')\n",
    "        return\n",
    "\n",
    "    result = ', '.join(str(random.randint(1, limit)) for r in range(rolls))\n",
    "    await ctx.send(result)\n",
    "\n",
    "\n",
    "@bot.command(description='For when you wanna settle the score some other way')\n",
    "async def choose(ctx, *choices: str):\n",
    "    \"\"\"Chooses between multiple choices.\"\"\"\n",
    "    await ctx.send(random.choice(choices))\n",
    "\n",
    "\n",
    "@bot.command()\n",
    "async def repeat(ctx, times: int, content='repeating...'):\n",
    "    \"\"\"Repeats a message multiple times.\"\"\"\n",
    "    for i in range(times):\n",
    "        await ctx.send(content)\n",
    "\n",
    "\n",
    "@bot.command()\n",
    "async def joined(ctx, member: discord.Member):\n",
    "    \"\"\"Says when a member joined.\"\"\"\n",
    "    await ctx.send(f'{member.name} joined {discord.utils.format_dt(member.joined_at)}')\n",
    "\n",
    "\n",
    "@bot.group()\n",
    "async def cool(ctx):\n",
    "    \"\"\"Says if a user is cool.\n",
    "\n",
    "    In reality this just checks if a subcommand is being invoked.\n",
    "    \"\"\"\n",
    "    if ctx.invoked_subcommand is None:\n",
    "        await ctx.send(f'No, {ctx.subcommand_passed} is not cool')\n",
    "\n",
    "\n",
    "@cool.command(name='bot')\n",
    "async def _bot(ctx):\n",
    "    \"\"\"Is the bot cool?\"\"\"\n",
    "    await ctx.send('Yes, the bot is cool.')\n",
    "\n",
    "\n",
    "bot.run('MTA4NTkyNzY0ODQzMTE5MDA3Nw.G4bEON.9uXJK4mnus4E95-_JdFYShIJRlq_pfCx_788Cs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2529134d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
